UserSearch: Users role added to search and filter

nuromirg opened the pull request:
**What is this feature?**

In this feature, we expand the search on the user page to roles, in addition to login, email and name, we can quickly find a given role. A new component has also been added that allows user to specify one or more roles to filter users by them.

![2024-08-26_13-09](https://github.com/user-attachments/assets/9d2d36b0-dea6-4f6d-b4c3-a6f239765f64)
![usersearch](https://github.com/user-attachments/assets/598bd406-f5d2-4cd6-8113-8b4af788ea4b)

The screenshots above show an example of what the component looks like and the filtering/searching result on basic roles.

**Why do we need this feature?**

In case we have many organizations and we need to be able to conveniently filter them by roles.

**Who is this feature for?**

For every user with search permissions.

**Which issue(s) does this PR fix?**:

Fixes [#92272](https://github.com/grafana/grafana/issues/92272)

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

nuromirg committed:
users role search/filter impelemented

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=92401) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=92401) before we can accept your contribution.<br/><hr/>**nuromirg** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=92401) it.</sub>

nuromirg committed:
aggregated unit tests for both filters

IevaVasiljeva reviewed (commented):
Thank you very much for the contribution! Filtering by organization roles is a feature that we'd really like to have.

The only big change I would make is adding this filter to the `Organization users` tab instead of the `All users` tab. This is mainly because this currently doesn't work nicely if a user is a member of several organizations, especially if they also have different organization roles in different orgs. In that case user's listed role/organization will change depending on whether the filter is applied, and whether it's ascending or descending.

Adding the feature to `Organization users` tab would also make it usable by org Admins and not just Server Admins (org Admins can't see `All users` tab). Moreover, it would also be usable in our Cloud instances (they are limited to one org only, so we only show the `Organization users` tab there).

The downside is that it would make the feature slightly less powerful (you'd need to switch between orgs to see the admins in different orgs). But I feel like we need a larger `All users` tab redesign if we want to list user orgs and roles there.

What do you think?

nuromirg commented on pkg/services/user/userimpl/store.go:
```diff
@@ -490,8 +494,8 @@ func (ss *sqlStore) Search(ctx context.Context, query *user.SearchUsersQuery) (*
 		if query.Query != "" {
-			whereConditions = append(whereConditions, "(email "+ss.dialect.LikeStr()+" ? OR name "+ss.dialect.LikeStr()+" ? OR login "+ss.dialect.LikeStr()+" ?)")
-			whereParams = append(whereParams, queryWithWildcards, queryWithWildcards, queryWithWildcards)
+			whereConditions = append(whereConditions, "(u.email "+ss.dialect.LikeStr()+" ? OR u.name "+ss.dialect.LikeStr()+" ? OR u.login "+ss.dialect.LikeStr()+" ? OR org_user.role "+ss.dialect.LikeStr()+" ?)")
+			whereParams = append(whereParams, queryWithWildcards, queryWithWildcards, queryWithWildcards, queryWithWildcards)
```
Thank you for your comments! I agree that it can be quite difficult for an administrator to work with the case when one user belongs to several organizations. This requires a reworking of the user list displaying itself, not only a search and filtering.

In this case, I can remove the current changes from `All Users`, they will remain as is, and for searching by organization I could add a filter for the role, but leave the search as it is now (for login, name and email). In this way, we will, firstly, be able to localize the search for the user by the combination of login details and role, and secondly, we will provide an uncluttered search for all possible user columns in the UI.

nuromirg committed:
org users role filter implemented (srv + ui)

nuromirg commented:
Greetings @IevaVasiljeva! I moved the role selection as a filter to the organization users section. Left the admin section as before, removing the role filter and role search. Also corrected the test cases, added two cases (on server side) for searching among the organization users.

![image](https://github.com/user-attachments/assets/e08a4be9-39c8-4fc8-ae45-42e99e94efff)

![image](https://github.com/user-attachments/assets/5f402382-9877-4b8e-82c0-27c5eb84407e)

Admin section remains the same

![image](https://github.com/user-attachments/assets/a5420fd5-acf4-409f-a341-c7d7e3b8d1ef)

If any additions to the current implementation are needed, I will be glad to hear them

nuromirg commented:
Hi @IevaVasiljeva, is this pull request no longer relevant? Can I close it?
